packages/core/saved-objects/core-saved-objects-migration-server-internal/src/actions/delete_by_query.ts

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import * as Either from 'fp-ts/lib/Either';
+import * as TaskEither from 'fp-ts/lib/TaskEither';
+import type {
+  BulkIndexByScrollFailure,
+  DeleteByQueryResponse,
+  QueryDslQueryContainer,
+} from '@elastic/elasticsearch/lib/api/types';
+import type { ElasticsearchClient } from '@kbn/core-elasticsearch-server';
+import {
+  catchRetryableEsClientErrors,
+  type RetryableEsClientError,
+} from './catch_retryable_es_client_errors';
+
+/** @internal */
+export interface DeleteByQueryParams {
+  client: ElasticsearchClient;
+  indexName: string;
+  query: QueryDslQueryContainer;
+}
+
+/** @internal */
+export interface DeleteByQueryErrorResponse {
+  type: 'delete_failed';
+  conflictingDocuments: BulkIndexByScrollFailure[];
+}
+
+/**
+ * Deletes documents matching the provided query
+ */
+export const deleteByQuery = ({
+  client,
+  indexName,
+  query,
+}: DeleteByQueryParams): TaskEither.TaskEither<
+  RetryableEsClientError | DeleteByQueryErrorResponse,
+  'delete_successful'
+> => {
+  return () => {
+    return client
+      .deleteByQuery({
+        index: indexName,
+        query,
+        wait_for_completion: true,
+        refresh: true,
```

Comment:
nit: can we add a comment why this is necessary because in the future someone might think "we don't need to force the refresh"